QUICKSTEP-57 Support for performing partitioned aggregation.

hbdeshmukh opened the pull request:
- Used for creating a pool of hash tables such that each hash table
  belongs to a unique partition.
- The partitioning is done on the group-by keys.
- Wrote a utility function to compute composite hash of a group of
  TypedValues.
- The Tuple class now supports a method to compute the hash of the entire
  tuple (i.e. hash key is the composite key made up of all the
  attributes in the tuple).
- Added a check for whether the aggregation is partitioned or not.
- The conditions for whether the aggregation can be partitioned
  are as follows:
  1. The query has a GROUP BY clause.
  2. There are no aggrgeations with a DISTINCT clause.
  3. The estimated number of groups are greater than a pre-defined
  threshold.
  4. The query has at least one aggregation function.
- Method for partitioned aggregation with GROUP BY
- StorageBlock now provides a method for performing GROUP BY aggregation
  in a partitioned way.
- AggregationOperationState calls appropriate method (i.e.
  aggregateGroupBy or aggregateGroupByPartitioned) based on the way in
  which aggregation is being performed.

Performance comparison: TPC-H SF100. 

**Query**|**master**|**partitioned aggregation**|**Speedup over master**
:-----:|:-----:|:-----:|:-----:
1|16,046|14,625|1.10
2|5,625|3,316|1.70
3|6,861|7,987|0.86
4|2,662|5,149|0.52
5|4,364|4,586|0.95
6|398|397|1.00
7|23,367|24,600|0.95
8|3,274|3,701|0.88
9|10,050|13,029|0.77
10|15,296|9,502|1.61
11|2,110|931|2.27
12|1,805|3,206|0.56
13|34,220|17,551|1.95
14|771|815|0.95
15|4,435|915|4.84
16|8,661|8,489|1.02
17|160,707|18,885|8.51
18|66,309|27,744|2.39
19|1,475|1,566|0.94
20|55,381|11,429|4.85
21|121,310|127,343|0.95
22|6,792|6,683|1.02
Total|551,921|312,448|1.77

Some queries e.g. 13, 15, 17, 18, 20 see drastic improvement, primarily because the FinalizeAggregation operation is multi-threaded, compared to the current single threaded implementation. Overall, we see a 1.77x improvement on the entire workload. 

For some queries we see significant degradation e.g. 4, 9 and 12. I looked at 9 and 12's individual runs and it seemed mostly an issue of large variance between the runs. For Q4, I saw a consistent degradation. So far, I couldn't figure out its reason.

hakanmemisoglu commented:
@hbdeshmukh, It seems OK to me. 
I guess Cmake just complains about a circular dependency.
You can remove `quickstep_storage_PartitionedHashTable` library from the CMakeList definition to break circular dependency.

hbdeshmukh commented:
Thanks @hakanmemisoglu 

I think the circular dependency is because of the inclusion of PartitionedHashTablePool in Storage Manager. If I cut that link, it should be fine. I may have to rewrite a function in StorageManager, which shouldn't be difficult.

hbdeshmukh commented on relational_operators/FinalizeAggregationOperator.cpp:
```diff
@@ -41,12 +41,29 @@ bool FinalizeAggregationOperator::getAllWorkOrders(
-            query_context->getAggregationState(aggr_state_index_),
-            query_context->getInsertDestination(output_destination_index_)),
-        op_index_);
+    DCHECK(query_context->getAggregationState(aggr_state_index_) != nullptr);
```
Reuse ``query_context->getAggregationState(aggr_state_index_)`` as a const pointer in the code below.

hbdeshmukh commented on storage/AggregationOperationState.hpp:
```diff
@@ -172,12 +173,44 @@ class AggregationOperationState {
 
  private:
+  static constexpr std::size_t kPartitionedAggregateThreshold = 100;
+  static constexpr std::size_t kNumPartitionsForAggregate = 40;
```
Convert these two constants to GFLAG arguments. That will allow us to experiment with an optimal configuration.

hbdeshmukh commented on relational_operators/FinalizeAggregationOperator.cpp:
```diff
@@ -41,12 +41,29 @@ bool FinalizeAggregationOperator::getAllWorkOrders(
-            query_context->getInsertDestination(output_destination_index_)),
-        op_index_);
+    DCHECK(query_context->getAggregationState(aggr_state_index_) != nullptr);
+    if (query_context->getAggregationState(aggr_state_index_)->isAggregatePartitioned()) {
```
We can unify the for loop for both the cases - partitioned and non-partitioned cases.

hbdeshmukh commented on storage/PartitionedHashTablePool.hpp:
```diff
@@ -0,0 +1,224 @@
+  inline std::size_t setHashTableSize(const std::size_t overall_estimate,
+                                      const std::size_t num_partitions) const {
+    CHECK_NE(num_partitions, 0Lu);
+    return std::max(static_cast<std::size_t>(overall_estimate/num_partitions), 26000Lu);
```
We can set the minimum to 100 and add an explanation here.

hbdeshmukh commented on storage/PartitionedHashTablePool.hpp:
```diff
@@ -0,0 +1,224 @@
+   * @param agg_handle The aggregation handle.
+   * @param storage_manager A pointer to the storage manager.
+   *
+   * @note The estimate of number of entries is quite inaccurate at this time.
```
This comment is not applicable, remove it.

hbdeshmukh commented on storage/PartitionedHashTablePool.hpp:
```diff
@@ -0,0 +1,224 @@
+   *
+   * @return A hash table pointer for the given HashTable.
+   **/
+  AggregationStateHashTableBase* getHashTableFast(const std::size_t partition_id) {
```
This function may not be needed.

hbdeshmukh commented:
Addressed the review comments and re-ran the queries. Here are the results: 

<table class="table table-bordered table-hover table-condensed">
<thead><tr><th title="Field #1">Queries</th>
<th title="Field #2">master oct 26</th>
<th title="Field #3">part-agg oct 26</th>
<th title="Field #4">Speedup over master</th>
</tr></thead>
<tbody><tr><td>1</td>
<td>16,904</td>
<td>16,295</td>
<td align="right">1.04</td>
</tr>
<tr><td>2</td>
<td>766</td>
<td>612</td>
<td align="right">1.25</td>
</tr>
<tr><td>3</td>
<td>5,060</td>
<td>5,105</td>
<td align="right">0.99</td>
</tr>
<tr><td>4</td>
<td>2,332</td>
<td>2,355</td>
<td align="right">0.99</td>
</tr>
<tr><td>5</td>
<td>4,988</td>
<td>5,025</td>
<td align="right">0.99</td>
</tr>
<tr><td>6</td>
<td>402</td>
<td>401</td>
<td align="right">1.00</td>
</tr>
<tr><td>7</td>
<td>28,398</td>
<td>24,098</td>
<td align="right">1.18</td>
</tr>
<tr><td>8</td>
<td>1,818</td>
<td>1,835</td>
<td align="right">0.99</td>
</tr>
<tr><td>9</td>
<td>8,858</td>
<td>12,839</td>
<td align="right">0.69</td>
</tr>
<tr><td>10</td>
<td>13,388</td>
<td>8,126</td>
<td align="right">1.65</td>
</tr>
<tr><td>11</td>
<td>2,321</td>
<td>998</td>
<td align="right">2.32</td>
</tr>
<tr><td>12</td>
<td>2,606</td>
<td>2,531</td>
<td align="right">1.03</td>
</tr>
<tr><td>13</td>
<td>34,860</td>
<td>17,832</td>
<td align="right">1.95</td>
</tr>
<tr><td>14</td>
<td>818</td>
<td>813</td>
<td align="right">1.01</td>
</tr>
<tr><td>15</td>
<td>4,656</td>
<td>900</td>
<td align="right">5.17</td>
</tr>
<tr><td>16</td>
<td>9,411</td>
<td>8,574</td>
<td align="right">1.10</td>
</tr>
<tr><td>17</td>
<td>1,693</td>
<td>1,732</td>
<td align="right">0.98</td>
</tr>
<tr><td>18</td>
<td>92,072</td>
<td>27,665</td>
<td align="right">3.33</td>
</tr>
<tr><td>19</td>
<td>1,430</td>
<td>1,473</td>
<td align="right">0.97</td>
</tr>
<tr><td>20</td>
<td>1,594</td>
<td>1,947</td>
<td align="right">0.82</td>
</tr>
<tr><td>21</td>
<td>139,356</td>
<td>178,787</td>
<td align="right">0.78</td>
</tr>
<tr><td>22</td>
<td>6,895</td>
<td>6,905</td>
<td align="right">1.00</td>
</tr>
<tr><td>Total</td>
<td>380,625</td>
<td>326,850</td>
<td align="right">1.16</td>
</tr>
</tbody></table>

jianqiao commented:
LGTM! Merging.
